File: examples/java/kora-java-resilient/src/main/java/io/koraframework/example/resilient/ComboService.java
package io.koraframework.example.resilient;

import java.util.concurrent.ThreadLocalRandom;
import io.koraframework.common.annotation.Component;
import io.koraframework.common.annotation.Root;
import io.koraframework.resilient.circuitbreaker.annotation.CircuitBreakable;
import io.koraframework.resilient.fallback.annotation.Fallback;
import io.koraframework.resilient.retry.annotation.Retryable;
import io.koraframework.resilient.timeout.annotation.Timeout;

@Root
@Component
public class ComboService {

    public static final String VALUE = "OK";
    public static final String FALLBACK = "FALLBACK";

    @Fallback(method = "getFallback()")
    @CircuitBreakable(MyCircuitBreaker.class)
    @Retryable(DefaultRetry.class)
    @Timeout(DefaultTimeouter.class)
    public String getValue(boolean fail) {
        if (fail) {
            throw new IllegalStateException("Failed");
        }

        try {
            int emulateWork = ThreadLocalRandom.current().nextInt(100, 1000);
            Thread.sleep(emulateWork);
            return VALUE;
        } catch (InterruptedException e) {
            throw new IllegalStateException(e);
        }
    }

    protected String getFallback() {
        return FALLBACK;
    }
}

File: examples/kotlin/kora-kotlin-resilient/src/main/kotlin/io/koraframework/kotlin/example/resilient/ComboService.kt
package io.koraframework.kotlin.example.resilient

import io.koraframework.common.annotation.Component
import io.koraframework.common.annotation.Root
import io.koraframework.resilient.circuitbreaker.annotation.CircuitBreakable
import io.koraframework.resilient.fallback.annotation.Fallback
import io.koraframework.resilient.retry.annotation.Retryable
import io.koraframework.resilient.timeout.annotation.Timeout
import java.util.concurrent.ThreadLocalRandom

@Root
@Component
open class ComboService {
    @Fallback(method = "getFallback()")
    @CircuitBreakable(MyCircuitBreaker::class)
    @Retryable(DefaultRetry::class)
    @Timeout(DefaultTimeouter::class)
    open fun getValue(fail: Boolean): String {
        if (fail) {
            throw IllegalStateException("Failed")
        }

        try {
            val emulateWork = ThreadLocalRandom.current().nextInt(100, 1000)
            Thread.sleep(emulateWork.toLong())
            return VALUE
        } catch (e: InterruptedException) {
            throw IllegalStateException(e)
        }
    }

    protected open fun getFallback(): String = FALLBACK

    companion object {
        const val VALUE = "OK"
        const val FALLBACK = "FALLBACK"
    }
}
